src/trs/hmtrs.rs

```diff
@@ -34,144 +92,106 @@ pub struct HMTRS {
     pub ctx: Context,
 }
 impl HMTRS {
-    /// the size of the HMTRS (the sum over the size of the rules in the [`TRS`])
+    /// The size of the HMTRS (the sum over the size of the rules in the [`TRS`])
     ///
     /// [`TRS`]: ../../term_rewriting/struct.TRS.html
     pub fn size(&self) -> usize {
         self.trs.size()
     }
-    /// Convert the `HMTRS` to a `String`.
-    pub fn display(&self) -> String {
-        self.trs.display(&self.signature)
-    }
-    /// All the free variables in the lexicon.
-    pub fn free_vars(&self) -> Vec<pVar> {
+    /// All the free type variables in the lexicon.
+    pub fn free_vars(&self) -> Vec<TypeVar> {
         let vars_fvs = self.vars.iter().flat_map(|x| x.free_vars());
         let ops_fvs = self.ops.iter().flat_map(|x| x.free_vars());
         vars_fvs.chain(ops_fvs).unique().collect()
     }
-    /// Infer the `Type` of a `Variable`.
-    ///
-    /// No [`Context`] is necessary for inference here; `self` either knows the correct [`TypeSchema`] or generates a [`TypeError`].
-    ///
-    /// [`Context`]: ../../polytype/struct.Context.html
-    /// [`TypeSchema`]: ../../polytype/struct.TypeSchema.html
-    /// [`TypeError`]: struct.TypeError.html
-    pub fn infer_var(&self, v: &tVar) -> Result<TypeSchema, TypeError> {
+    fn var_tp(&self, v: &Variable) -> Result<TypeSchema, TypeError> {
         if let Some(idx) = self.signature.variables().iter().position(|x| x == v) {
             Ok(self.vars[idx].clone())
         } else {
-            Err(TypeError)
+            Err(TypeError::VarNotFound)
         }
     }
-    /// Infer the `Type` of a `Variable`.
-    ///
-    /// No [`Context`] is necessary for inference here; `self` either knows the correct [`TypeSchema`] or generates a [`TypeError`].
-    ///
-    /// [`Context`]: ../../polytype/struct.Context.html
-    /// [`TypeSchema`]: ../../polytype/struct.TypeSchema.html
-    /// [`TypeError`]: struct.TypeError.html
-    pub fn infer_op(&self, o: &Operator) -> Result<TypeSchema, TypeError> {
+    fn op_tp(&self, o: &Operator) -> Result<TypeSchema, TypeError> {
         if let Some(idx) = self.signature.operators().iter().position(|x| x == o) {
             Ok(self.ops[idx].clone())
         } else {
-            Err(TypeError)
+            Err(TypeError::OpNotFound)
         }
     }
-    /// Infer the [`TypeSchema`] of a [`Term`] and the corresponding [`Context`] or generate a [`TypeError`].
+    /// Infer the [`TypeSchema`] of a [`Term`].
```

Comment:
Hmm... why not tell people that the context is also inferred?

Reply:
I deleted this text because it follows from the spec (taking a mutable context, returning a `Result<_, TypeError>`)